From 38908c7d38b51339bea0b7189e0aae1116530914 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 11 Mar 2025 14:13:11 +0100
Subject: [PATCH 1636/1695] clk: error when set_rate ops is not supported

The function clk_set_rate() returns the error TEE_ERROR_NOT_SUPPORTED
if this operant is not supported by the clock, when the flag
CLK_SET_RATE_PARENT is absent and when the ops set_rate is absent.

This patch allows the caller to detect an issue when set_rate ops is not
performed without need to request again clk_get_rate().

clk_compute_rate_no_lock() is still called to recompute frequency as parent
rate can change.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0aa89449ef848d0cc28b667f4facaac3d8aa770e clk: error when set_rate ops is not supported]
Fixes: 79591a69e1e5 ("clk: add CLK_SET_RATE_PARENT flags")
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I62b83a7e41a0d3a68bda164eca70af5e7706bdde
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/443430
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/clk/clk.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/core/drivers/clk/clk.c b/core/drivers/clk/clk.c
index 718b2f636..d801aac36 100644
--- a/core/drivers/clk/clk.c
+++ b/core/drivers/clk/clk.c
@@ -58,6 +58,9 @@ static bool __maybe_unused clk_check(struct clk *clk)
 	if (clk->num_parents > 1 && !clk->ops->get_parent)
 		return false;
 
+	if (clk->ops->determine_rate && !clk->ops->set_rate)
+		return false;
+
 	return true;
 }
 
@@ -360,6 +363,7 @@ static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
 	unsigned long parent_rate = 0;
 	unsigned long new_rate = rate;
 	struct clk *parent = clk->parent;
+	bool skip_set_rate = false;
 
 	if (parent)
 		parent_rate = parent->rate;
@@ -407,6 +411,7 @@ static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
 		if (res)
 			return res;
 		new_rate = parent->rate;
+		skip_set_rate = true;
 	}
 
 	if (clk->ops->set_rate) {
@@ -420,6 +425,8 @@ static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
 
 		if (res)
 			return res;
+	} else if (!skip_set_rate) {
+		res = TEE_ERROR_NOT_SUPPORTED;
 	}
 
 	clk_compute_rate_no_lock(clk);
-- 
2.47.2

